sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/resolution/Resolver.kt

```diff
@@ -0,0 +1,220 @@
+ *     the top most result set will be passed to the compiler to create that mapper.
+ *
+ * The important distinction is that the resolver is stateful and validators are stateless.
+ * The resolver modifies its scope while running and uses that to validate SQL.
+ *
+ * Both the resolver and the validator can return different types of errors. Validators will always
+ * return errors that are independent of state (such as using the DISTINCT keyword in a
+ * non-aggregate function). Resolvers will always return errors that depend on state (trying to
+ * return a result column that doesnt exist).
+ *
+ */
+data class Resolver(
+    /**
+     * Symbol Table containing all the available tables this resolver can select from.
+     */
+    internal val symbolTable: SymbolTable,
+
+    /**
+     * Dependencies are a way of communicating to the caller which table tags were used in
+     * resolution. In the IntelliJ plugin, table tags are virtual files, so this translates to
+     * "which files were used in resolution".
+     */
+    internal val dependencies: LinkedHashSet<Any> = linkedSetOf<Any>(),
+
+    /**
+     * Values that have been provided by a parent Resolver. This is exclusively used when
+     * resolving as part of an expression. Example:
+     *
+     * SELECT *
+     * FROM test AS test1
+     * WHERE 'sup' IN (
+     *   SELECT *
+     *   FROM test AS test2
+     *   WHERE test2.some_id = test1.some_id
+     * );
+     *
+     * In this example 'test' resolves to 'test.some_id'. Normally resolving the
+     * expression select statement would fail with "no table named test1" but because
+     * we scope in the current resolution (in this case the resolution of "test" aliased as test1)
+     * it is able to resolve "test1.some_id".
+     */
+    internal val scopedValues: List<Value> = emptyList(),
+
+    /**
+     * The offset of an element we wish to know the source of. Note that if we ever change contexts
+     * as part of resolution (for example if we need to resolve a view as part of a select's
+     * resolution) then this cannot be passed to the resolver being used for the new context.
+     * This gets used by the IntelliJ plugin to perform its own reference resolution.
+     */
+    internal val elementToFind: Int? = null,
+
+    /**
+     * Used to prevent cyclic view resolution.
+     */
+    private val currentlyResolvingViews: LinkedHashSet<String> = linkedSetOf<String>()
+) {
+
+  /**
+   * @return a resolver with the with statement's tables included in the symbol table.
+   */
+  internal fun withResolver(with: SqliteParser.With_clauseContext) =
+      copy(with.common_table_expression().fold(symbolTable, { symbolTable, commonTable ->
+        symbolTable + SymbolTable(commonTable, commonTable)
+      }))
+
+  /**
+   * Take a table or subquery rule and return a list of the selected values.
+   */
+  internal fun resolve(tableOrSubquery: SqliteParser.Table_or_subqueryContext): Resolution {
```

Comment:
originally my thought was "well all of the other resolvers need to resolve tables/columns so that should be on the type" but those methods call into the other resolvers so the sentiment doesnt really make sense. I'll put it into its own "TableResolver"